llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -115,6 +115,25 @@ bool VPRecipeBase::mayWriteToMemory() const {
 
 bool VPRecipeBase::mayReadFromMemory() const {
   switch (getVPDefID()) {
+  case VPInstructionSC:
+    if (Instruction::isBinaryOp(cast<VPInstruction>(this)->getOpcode()))
+      return false;
+    switch (cast<VPInstruction>(this)->getOpcode()) {
+    case Instruction::Or:
+    case Instruction::ICmp:
+    case Instruction::Select:
+    case VPInstruction::AnyOf:
+    case VPInstruction::Not:
+    case VPInstruction::CalculateTripCountMinusVF:
+    case VPInstruction::CanonicalIVIncrementForPart:
+    case VPInstruction::ExtractFromEnd:
+    case VPInstruction::FirstOrderRecurrenceSplice:
+    case VPInstruction::LogicalAnd:
+    case VPInstruction::PtrAdd:
+      return false;
+    default:
+      return true;
+    }
```

Comment:
Those effectively check if the VPInstruction access memory (either reading or writing). Could you move it to something like `VPInstruction::mayReadOrWriteFromMemory()` and use it for both `VPRecipeBase::mayReadFromMemory` and `VPRecipeBase::mayWriteToMemory`, so it is automatically keept in sync?

Reply:
I'm not sure where to put this, since there's already a `VPRecipeBase::mayReadOrWriteMemory()` and it seems a bit hairy to override that. I could name it something like `VPInstruction::opcodeMayAccessMemory()` or something?

Alternatively, would these end up diverging at some point? Assuming we'll never have a VPInstruction with an ordered load or store we could mark those as free from writes and stores respectively. Or is that out of the scope of this PR?

Reply:
I've pulled it out into a static function to avoid exposing another interface for it, but happy to rework it however

Reply:
Hmmm, I think it should be part of `VPInstruction`, VPRecipeBase `mayReadFromMemory/mayWriteToMemory` can call it after casting to VPInstruction, and through those VPRecipeBase `mayReadOrWriteMemory` will also return the correct result?

> Alternatively, would these end up diverging at some point? Assuming we'll never have a VPInstruction with an ordered load or store we could mark those as free from writes and stores respectively. Or is that out of the scope of this PR?

Not sure what you mean with diverging? The VPInstruction version should only mark opcodes as non writing/reading if we are sure they won't access memory